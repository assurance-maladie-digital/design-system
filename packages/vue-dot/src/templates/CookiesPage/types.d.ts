import { TypeEnum } from './CookiesInformation/TypeEnum';
import { CookiesList } from './CookiesTable/types';

export type CookiesPageItems = {
	[key in TypeEnum]?: CookiesList[];
};

export type Preferences = {
	[key: string]: boolean | null;
	[TypeEnum.FUNCTIONAL]: boolean | null;
	[TypeEnum.ANALYTICS]: boolean | null;
};
