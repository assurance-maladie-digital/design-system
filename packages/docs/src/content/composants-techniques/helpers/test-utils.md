---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
