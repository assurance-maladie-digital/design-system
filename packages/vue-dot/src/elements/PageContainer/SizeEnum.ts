export enum SizeEnum {
	X_LARGE = 'xl',
	LARGE = 'l',
	MEDIUM = 'm',
	SMALL = 's'
}

export const SIZE_ENUM_VALUES = Object.values(SizeEnum);
