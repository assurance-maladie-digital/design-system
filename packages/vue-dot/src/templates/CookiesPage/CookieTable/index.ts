import CookieTable from './CookieTable.vue';

export default CookieTable;
