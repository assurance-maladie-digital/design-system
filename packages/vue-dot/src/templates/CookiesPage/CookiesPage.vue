<template>
	<PageContainer
		size="m"
		class="vd-cookies-page"
	>
		<VCard class="pa-4 pa-md-8">
			<VForm ref="form">
				<h2 class="text-h5 font-weight-bold mb-6">
					{{ locales.title }}
				</h2>

				<p>
					{{ locales.description }}
				</p>

				<p>
					{{ locales.cookieDefinition }}
				</p>

				<div class="d-flex flex-wrap justify-end mx-n2 mt-n2 mb-6">
					<VBtn
						color="primary"
						outlined
						class="ma-2"
						@click="setGlobalPreferences(false)"
					>
						{{ locales.rejectAll }}
					</VBtn>

					<VBtn
						color="primary"
						outlined
						class="ma-2"
						@click="setGlobalPreferences(true)"
					>
						{{ locales.acceptAll }}
					</VBtn>
				</div>

				<CookiesInformation
					v-for="(cookies, key) in items"
					:key="key"
					v-model="preferences[key]"
					:type="key"
					:table-items="cookies"
					class="mb-6"
				/>

				<div class="d-flex mt-16">
					<VSpacer />

					<VBtn
						color="primary"
						@click="submit"
					>
						{{ locales.saveBtn }}
					</VBtn>
				</div>
			</VForm>
		</VCard>
	</PageContainer>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { locales } from './locales';
	import { CookiesPageItems, Preferences } from './types';

	import CookiesInformation from './CookiesInformation';

	import { Refs, VForm } from '../../types';
	import { TypeEnum } from './CookiesInformation/TypeEnum';

	const Props = Vue.extend({
		props: {
			items: {
				type: Object as PropType<CookiesPageItems>,
				required: true
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component({
		components: {
			CookiesInformation
		}
	})
	export default class CookiesPage extends MixinsDeclaration {
		$refs!: Refs<{
			form: VForm;
		}>;

		locales = locales;

		preferences: Preferences = {
			[TypeEnum.FUNCTIONAL]: null,
			[TypeEnum.ANALYTICS]: null
		};

		get filteredPreferences(): Partial<Preferences> {
			const filtered: Partial<Preferences> = {};

			Object.entries(this.preferences).forEach(([key, value]) => {
				if (this.items[key as TypeEnum] !== undefined) {
					filtered[key] = value;
				}
			});

			return filtered;
		}

		setGlobalPreferences(value: boolean): void {
			this.preferences[TypeEnum.FUNCTIONAL] = value;
			this.preferences[TypeEnum.ANALYTICS] = value;
		}

		async submit(): Promise<void> {
			await this.$nextTick();

			if (!this.$refs.form.validate()) {
				return;
			}

			this.$emit('submit', this.filteredPreferences);
		}
	}
</script>
