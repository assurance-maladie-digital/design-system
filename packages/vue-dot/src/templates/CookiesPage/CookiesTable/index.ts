import CookiesTable from './CookiesTable.vue';

export default CookiesTable;
