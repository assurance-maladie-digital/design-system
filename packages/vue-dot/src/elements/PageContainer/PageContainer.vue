<template>
	<div
		:class="spacingClass"
		class="vd-page-container d-flex justify-center"
	>
		<VSheet
			:width="containerSize"
			:color="color"
		>
			<slot />
		</VSheet>
	</div>
</template>

<script lang="ts">
	import Vue from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { SizeEnum, SIZE_ENUM_VALUES } from './SizeEnum';

	import { IndexedObject } from '../../types';

	import { propValidator } from '../../helpers/propValidator';

	const Props = Vue.extend({
		props: {
			size: {
				type: String,
				default: SizeEnum.X_LARGE,
				validator: (value: string) => propValidator('size', SIZE_ENUM_VALUES, value)
			},
			spacing: {
				type: String,
				default: undefined
			},
			color: {
				type: String,
				default: 'transparent'
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component
	export default class PageContainer extends MixinsDeclaration {
		get spacingClass(): string {
			if (this.spacing) {
				return this.spacing;
			}

			const spacingMapping: IndexedObject<string> = {
				xs: 'px-0',
				sm: 'px-4',
				md: 'px-8',
				lg: 'px-8',
				xl: 'px-8'
			};

			const spacing = spacingMapping[this.$vuetify.breakpoint.name];

			return `py-10 ${spacing}`;
		}

		get containerSize(): number {
			const sizeMapping: IndexedObject<number> = {
				[SizeEnum.X_LARGE]: 1440,
				[SizeEnum.LARGE]: 960,
				[SizeEnum.MEDIUM]: 800,
				[SizeEnum.SMALL]: 600
			};

			return sizeMapping[this.size];
		}
	}
</script>
