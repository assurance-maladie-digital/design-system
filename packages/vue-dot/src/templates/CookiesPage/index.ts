import CookiesPage from './CookiesPage.vue';

export default CookiesPage;
