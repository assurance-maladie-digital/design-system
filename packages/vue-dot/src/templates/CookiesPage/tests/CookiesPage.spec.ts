import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import CookiesPage from '../';
import PageContainer from '../../../elements/PageContainer';

import { cookiesList } from './data/cookiesList';

Vue.component('PageContainer', PageContainer);

let wrapper: Wrapper<Vue>;

describe('CookiesPage', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(CookiesPage, {
			propsData: {
				items: cookiesList
			}
		});

		expect(wrapper).toMatchSnapshot();
	});
});
