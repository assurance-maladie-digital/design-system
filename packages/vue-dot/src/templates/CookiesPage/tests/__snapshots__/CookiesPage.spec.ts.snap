// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookiesPage renders correctly 1`] = `
<pagecontainer-stub size="m" color="transparent" class="vd-cookies-page">
  <vcard-stub loaderheight="4" tag="div" class="pa-4 pa-md-8">
    <vform-stub>
      <h2 class="text-h5 font-weight-bold mb-6">
        Gestion des cookies
      </h2>
      <p>
        Définissez vos préférences de collecte et de traitement de vos données sur ce site.
      </p>
      <p>
        Un cookie est un petit fichier texte déposé sur votre terminal (ordinateur, tablette ou mobile) lors de la visite d’un site ou lors de la consultation d’une publicité. Il contient plusieurs données&nbsp;: le nom du service qui l’a déposé, un identifiant sous forme de numéro unique, une valeur et éventuellement une date d’expiration.
      </p>
      <div class="d-flex flex-wrap justify-end mx-n2 mt-n2 mb-6">
        <vbtn-stub color="primary" outlined="true" tag="button" activeclass="" type="button" class="ma-2">
          Tout refuser
        </vbtn-stub>
        <vbtn-stub color="primary" outlined="true" tag="button" activeclass="" type="button" class="ma-2">
          Tout accepter
        </vbtn-stub>
      </div>
      <cookiesinformation-stub type="essentials" tableitems="[object Object],[object Object]" class="mb-6"></cookiesinformation-stub>
      <cookiesinformation-stub type="functional" tableitems="[object Object],[object Object]" class="mb-6"></cookiesinformation-stub>
      <cookiesinformation-stub type="analytics" tableitems="[object Object],[object Object]" class="mb-6"></cookiesinformation-stub>
      <div class="d-flex mt-16">
        <vspacer-stub tag="div"></vspacer-stub>
        <vbtn-stub color="primary" tag="button" activeclass="" type="button">
          Enregistrer mes choix
        </vbtn-stub>
      </div>
    </vform-stub>
  </vcard-stub>
</pagecontainer-stub>
`;
